File: lib/billing/email.ts
import { Resend } from "resend";
import { appConfig } from "@/lib/app-config";

const RESEND_API_KEY = process.env.RESEND_API_KEY;
const FROM_EMAIL = process.env.BILLING_FROM_EMAIL ?? appConfig.billingEmail;

export const isEmailEnabled = !!RESEND_API_KEY;
const resend = isEmailEnabled ? new Resend(RESEND_API_KEY) : null;

export async function sendBillingEmail(params: {
  to: string;
  subject: string;
  html: string;
}): Promise<boolean> {
  if (!resend) {
    console.log(`[Email Stub] To: ${params.to} | Subject: ${params.subject}`);
    return false;
  }

  try {
    await resend.emails.send({
      from: FROM_EMAIL,
      to: params.to,
      subject: params.subject,
      html: params.html,
    });
    return true;
  } catch (err) {
    console.error("Email send failed:", err);
    return false;
  }
}

File: rustbill/crates/rustbill-core/src/notifications/email.rs
//! Email sending via Resend HTTP API (matching existing integration).

use reqwest::Client;
use serde::Serialize;

#[derive(Debug, Clone)]
pub struct EmailSender {
    http: Client,
    api_key: String,
    from: String,
}

#[derive(Debug, Serialize)]
struct ResendRequest<'a> {
    from: &'a str,
    to: &'a [&'a str],
    subject: &'a str,
    html: &'a str,
}

impl EmailSender {
    pub fn new(api_key: String, from: String) -> Self {
        Self {
            http: Client::new(),
            api_key,
            from,
        }
    }

    /// Try to create from environment variables. Returns None if RESEND_API_KEY is not set.
    pub fn from_env() -> Option<Self> {
        let api_key = std::env::var("RESEND_API_KEY").ok()?;
        let from = std::env::var("BILLING_FROM_EMAIL")
            .unwrap_or_else(|_| "billing@rantai.com".to_string());
        Some(Self::new(api_key, from))
    }

    /// Send an email. Returns true on success, false on failure (non-blocking).
    pub async fn send(&self, to: &str, subject: &str, html: &str) -> bool {
        let body = ResendRequest {
            from: &self.from,
            to: &[to],
            subject,
            html,
        };

        match self
            .http
            .post("https://api.resend.com/emails")
            .header("Authorization", format!("Bearer {}", self.api_key))
            .json(&body)
            .send()
            .await
        {
            Ok(resp) if resp.status().is_success() => true,
            Ok(resp) => {
                tracing::warn!(status = %resp.status(), "Email send failed");
                false
            }
            Err(e) => {
                tracing::warn!(error = %e, "Email send error");
                false
            }
        }
    }
}
